### PR TITLE
Add GCC & Clang amd64 build CI

### PR DESCRIPTION
**Pull request by zefklop**
## Purpose

What the title says

## Proposed changes

- Fix build for clang-cl (amd64)
- Use build matrices to build Linux & MSVC builds
- Use github' cache to retrieve RosBE (for now using a custom build script, later with official RosBE)
  - Build for host architecture and use that as a cache key
  - Hash some toolchain specifics and feed that to ccache in order to not miss entries just because we changed host CPU
- Some improvements to be able to use llvm-rc (not there yet)
- Fix the ridiculous situation of framedyn.dll

## TODO

Maybe add Windows RosBE build.

### CODE REVIEW

File: dll/win32/kernel32/winnls/nls/afk.nls [hunk 1]
```diff
@@ -18,13 +18,11 @@
  * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
  */
 
-#pragma code_page(1252)
-
-LANGUAGE LANG_AFRIKAANS, SUBLANG_DEFAULT
-LOCALE_FONTSIGNATURE RCDATA { L"\x002f\x8000\x3808\x0000\x0000\x0000\x0000\x0000\x0001\x0000\x0000\x4000\x00ff\x003f\x0000\xffff" }
+#pragma code_page(65001) /* UTF-8 */
```
Comment (HBelusca) [inline dll/win32/kernel32/winnls/nls/afk.nls:21]: You may as well remove all these pragmas from these files, and "factor it out" into the main locale_rc.rc, as we do in our other translations.

File: dll/win32/kernel32/winnls/locale_rc.rc [hunk 2]
```diff
@@ -15,377 +15,248 @@
...
- * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
  */
 
-#include <windef.h>
-//#include "winbase.h"
-#include <winnls.h>
+#include "windef.h"
+#include "winbase.h"
+#include "winnls.h"
```
Comment (HBelusca) [inline dll/win32/kernel32/winnls/locale_rc.rc:23]: Why in quotes? These files are from the "main" headers (PSDK) -> should be better between chevrons?

File: dll/win32/framedyn/chstring.cpp [hunk 3]
```diff
@@ -348,30 +352,32 @@ void CHString::AllocCopy(CHString& dest, int nCopyLen, int nCopyIndex, int nExtr
...
+    wcsncpy(reinterpret_cast<LPWSTR>(dest.m_pchData),
+            reinterpret_cast<LPWSTR>(m_pchData + nCopyIndex),
+            nCopyLen);
 }
 
 /*
  * @implemented
  */
-BSTR CHString::AllocSysString() const
+CHSTRING_LPWSTR CHString::AllocSysString() const
```
Comment (HBelusca) [inline dll/win32/framedyn/chstring.cpp:363]: Why CHSTRING_LPWSTR? Is it equivalent to BSTR? (Note that https://docs.microsoft.com/en-us/windows/win32/api/chstring/nf-chstring-chstring-allocsysstring says it returns a BSTR.)

File: sdk/include/psdk/winnls.h [hunk 4]
```diff
@@ -116,42 +116,45 @@ extern "C" {
...
+#define LOCALE_INEGSYMPRECEDES      86
+#define LOCALE_INEGSEPBYSPACE       87
+#define LOCALE_FONTSIGNATURE        88
+#define LOCALE_SISO639LANGNAME      89
+#define LOCALE_SISO3166CTRYNAME     90
+
+/* FIXME: This value should be in the guarded block below */
+#define LOCALE_SNAME                92
+
+#if (WINVER >= 0x0600) || (defined(__REACTOS__) && defined(_KERNEL32_))
```
Comment (HBelusca) [inline sdk/include/psdk/winnls.h:132]: Why does this need a `__REACTOS__` guard?
(is it because we use, from the wine winnls code, these defines even if the kernel32 is compiled with WINVER == 0x0502 ?? 🤔 )

File: sdk/lib/crt/crt.cmake [hunk 5]
```diff
@@ -483,6 +483,10 @@ elseif(ARCH STREQUAL "amd64")
     if(MSVC)
         list(APPEND CRT_ASM_SOURCE
             except/amd64/cpp.s)
+        if (CMAKE_C_COMPILER_ID STREQUAL "Clang")
+            list(APPEND CRT_ASM_SOURCE
+                stdlib/clang-alias.s)
```
Comment (HBelusca) [inline sdk/lib/crt/crt.cmake:488]: Do we have this file? (I don't see it?)

File: sdk/tools/spec2def/spec2def.c [hunk 6]
```diff
@@ -712,8 +715,9 @@ OutputLine_def_GCC(FILE *fileDest, EXPORT *pexp)
         bTracing = 1;
     }
 
-    /* Special handling for stdcall and fastcall */
+    /* Special handling for stdcall and fastcall, but not C++ exports*/
```
Comment (HBelusca) [inline sdk/tools/spec2def/spec2def.c:718]:
```suggestion
    /* Special handling for stdcall and fastcall, but not C++ exports */
```

File: sdk/include/psdk/chstring.h [hunk 7]
```diff
@@ -6,125 +6,372 @@
...
 #include <provexce.h>
 
+/*
+ * Framedyn dates from ancient times when wchar_t was a typedef to unsigned short.
+ * In order be able to use newer compilers, we define inline overloaded wrappers.
+ */
+using CHSTRING_WCHAR=unsigned short ;
+using CHSTRING_LPCWSTR=const CHSTRING_WCHAR*;
+using CHSTRING_LPWSTR=CHSTRING_WCHAR*;
+static_assert(sizeof(CHSTRING_WCHAR) == sizeof(wchar_t), "CHSTRING_WCHAR must be of same size than wchar_t");
```
Comment (HBelusca) [inline sdk/include/psdk/chstring.h:16]:
```suggestion
typedef unsigned short CHSTRING_WCHAR;
typedef const CHSTRING_WCHAR *CHSTRING_LPCWSTR;
typedef CHSTRING_WCHAR *CHSTRING_LPWSTR;
C_ASSERT(sizeof(CHSTRING_WCHAR) == sizeof(wchar_t));
```

File: sdk/include/psdk/chstring.h [hunk 8]
```diff
@@ -6,125 +6,372 @@
...
-inline BOOL operator<=(const CHString& s1, LPCWSTR s2) { return s1.Compare(s2) <= 0; }
+inline BOOL operator<=(const CHString& s1, CHSTRING_LPCWSTR s2) { return s1.Compare(s2) <= 0; }
 inline BOOL operator<=(const CHString& s1, const CHString& s2) { return s1.Compare(s2) <= 0; }
 
-inline BOOL operator>=(const CHString& s1, LPCWSTR s2) { return s1.Compare(s2) >= 0; }
+inline BOOL operator>=(const CHString& s1, CHSTRING_LPCWSTR s2) { return s1.Compare(s2) >= 0; }
 inline BOOL operator>=(const CHString& s1, const CHString& s2) { return s1.Compare(s2) >= 0; }
 
+/* Have GCC link to the symbols exported by framedyn.dll */
+#ifdef __GNUC__
```
Comment (HBelusca) [inline sdk/include/psdk/chstring.h:199]: Why can't we hide this in the source code of framedyn itself as it previously was?

File: sdk/include/psdk/chstring.h (hunk 7 above)
Comment (zefklop) [reply sdk/include/psdk/chstring.h:16]: Why ?

File: sdk/lib/crt/crt.cmake (hunk 5 above)
Comment (binarymaster) [reply sdk/lib/crt/crt.cmake:488]: Yes we do: https://github.com/reactos/reactos/blob/master/sdk/lib/crt/stdlib/clang-alias.s

File: dll/win32/kernel32/winnls/nls/afk.nls (hunk 1 above)
Comment (zefklop) [reply dll/win32/kernel32/winnls/nls/afk.nls:21]: I just copy-pasted triste from wine and see no reason to introduce differences

File: dll/win32/kernel32/winnls/locale_rc.rc (hunk 2 above)
Comment (zefklop) [reply dll/win32/kernel32/winnls/locale_rc.rc:23]: Same answer as above

File: dll/win32/framedyn/chstring.cpp (hunk 3 above)
Comment (zefklop) [reply dll/win32/framedyn/chstring.cpp:363]: Mangled export says it returns a pointer to unsigned to unsigned short. So be it.

File: sdk/include/psdk/winnls.h (hunk 4 above)
Comment (zefklop) [reply sdk/include/psdk/winnls.h:132]: Yes

File: sdk/include/psdk/chstring.h (hunk 7 above)
Comment (zefklop) [reply sdk/include/psdk/chstring.h:16]: This is not an attempt to anything. And modern here mostly means "compilant to a 10-year-old standard". I think we settled this a while ago...

File: sdk/include/psdk/chstring.h (hunk 8 above)
Comment (zefklop) [reply sdk/include/psdk/chstring.h:199]: Because you might want to link your code to an import lib generated outside of the ReactOS tree.

**Comment by binarymaster** [comment]
Is it just me, or build times has increased significantly in this PR? :thinking: 

image

Build times from other PR for comparison:

image

**Comment by zefklop** [comment]
> Is it just me, or build times has increased significantly in this PR? 🤔
> 
> image
> 
> Build times from other PR for comparison:
> 
> image

It's because it compiled RosBE for the first time & the ccache cache is like new again. Relaunching it to prove this.

File: dll/win32/kernel32/winnls/nls/afk.nls (hunk 1 above)
Comment (HBelusca) [reply dll/win32/kernel32/winnls/nls/afk.nls:21]: ok so let's keep these like that.

File: .github/workflows/build.yml [hunk 9]
```diff
@@ -77,116 +77,58 @@ jobs:
...
+      matrix:
+        compiler: [msvc, clang-cl]
+        arch: [i386, amd64]
+      fail-fast: false
     runs-on: windows-latest
     steps:
-    - name: Install packages
+    - name: Install ninja
+      run: choco install -y ninja
+    - name: Install LLVM (x86)
```
Comment (SergeGautherie) [inline .github/workflows/build.yml:90]: Prefer
```suggestion
    - name: Install LLVM (i386)
```

File: .github/workflows/build.yml [hunk 10]
```diff
@@ -77,116 +77,58 @@ jobs:
...
+        echo "LLVM_PATH=${env:PROGRAMFILES}\llvm\bin" | Out-File -FilePath $env:GITHUB_ENV -Encoding utf8 -Append
+    - name: Install Flex & Bison
       run: |
         curl -O https://svn.reactos.org/storage/vperevertkin/flexbison.7z
         7z x flexbison.7z -O${{github.workspace}}\bin
         echo "${{github.workspace}}\bin" | Out-File -FilePath $env:GITHUB_PATH -Encoding utf8 -Append
         echo "BISON_PKGDATADIR=${{github.workspace}}\bin\share\bison" | Out-File -FilePath $env:GITHUB_ENV -Encoding utf8 -Append
         echo "M4=${{github.workspace}}\bin\m4.exe" | Out-File -FilePath $env:GITHUB_ENV -Encoding utf8 -Append
-    - name: Add CL to PATH
+    - name: Activate VS cmd (x86)
```
Comment (SergeGautherie) [inline .github/workflows/build.yml:107]: Prefer
```suggestion
    - name: Activate VS cmd (i386)
```

File: boot/freeldr/fdebug/lang/bg-BG.rc [hunk 11]
```diff
@@ -26,8 +26,8 @@ END
 /* Accelerators */
 IDC_FDEBUG ACCELERATORS
 BEGIN
-    63, IDM_ABOUT, ASCII, ALT // "?"
-    47, IDM_ABOUT, ASCII, ALT // "/"
+    VK_OEM_2, IDM_ABOUT, VIRTKEY, ALT //FIXME: ?/ US keyboard only
+    VK_DIVIDE, IDM_ABOUT, VIRTKEY, ALT
```
Comment (SergeGautherie) [inline boot/freeldr/fdebug/lang/bg-BG.rc:30]: It looks like freeldr files should be moved to a separate commit.

File: .github/workflows/build.yml [hunk 12]
```diff
@@ -2,107 +2,81 @@ name: Build
...
+    - name: Install ccache
       run: sudo apt install ccache
+    - name: Install clang
+      if: ${{matrix.compiler=='clang'}}
+      run: |
+        echo "LLVM_VERSION=12" >> $GITHUB_ENV
+        wget https://apt.llvm.org/llvm.sh
+        chmod +x llvm.sh
+        sudo ./llvm.sh $LLVM_VERSION
     - uses: actions/checkout@v2
```
Comment (SergeGautherie) [inline .github/workflows/build.yml:37]: Match
```suggestion
    - name: Source checkout
      uses: actions/checkout@v2
```

File: .github/workflows/build.yml [hunk 13]
```diff
@@ -2,107 +2,81 @@ name: Build
...
+        path: RosBE-CI
+        key: RosBE-CI-${{runner.os}}-${{steps.get_native_arch.outputs.march}}
+    - name: Compile RosBE
+      run: |
+        wget https://gist.githubusercontent.com/zefklop/b2d6a0b470c70183e93d5285a03f5899/raw/build_rosbe_ci.sh
+        chmod +x build_rosbe_ci.sh
+        ./build_rosbe_ci.sh ${{github.workspace}}/RosBE-CI
+    - name: Install ccache
       run: sudo apt install ccache
+    - name: Install clang
```
Comment (SergeGautherie) [inline .github/workflows/build.yml:30]: Match
```suggestion
    - name: Install LLVM
```

File: .github/workflows/build.yml [hunk 14]
```diff
@@ -2,107 +2,81 @@ name: Build
...
     - name: Set ccache settings
       run: |
         echo "CCACHE_BASEDIR=${{github.workspace}}" >> $GITHUB_ENV
         echo "CCACHE_DIR=${{github.workspace}}/ccache" >> $GITHUB_ENV
         echo "CCACHE_MAXSIZE=1G" >> $GITHUB_ENV
         echo "CCACHE_SLOPPINESS=time_macros" >> $GITHUB_ENV
+        echo "CCACHE_COMPILERCHECK=string:${{hashfiles('./build_rosbe_ci.sh', './RosBE-CI/git-commit')}}" >> $GITHUB_ENV
     - name: Configure
       run: |
         mkdir build
```
Comment (SergeGautherie) [inline .github/workflows/build.yml:56]: Match separate
```
    - name: Create Build directory
      run: mkdir build
```

File: .github/workflows/build.yml (hunk 9 above)
Comment (zefklop) [reply .github/workflows/build.yml:90]: Well, chocolatey says "--x86", let's respect that.

File: .github/workflows/build.yml (hunk 10 above)
Comment (zefklop) [reply .github/workflows/build.yml:107]: Same, MSVC prefers the x86 semantics

**Comment by SergeGautherie** [comment]
> Maybe add Windows RosBE build.

That would be good (At least @JoachimHenze uses that setup), MSBuild amd64 too.
They could be follow-up PRs...

**Review by Extravert-ir — APPROVED** [review]
